Add Troubleshoot Read and Write latency issue document (#3615)
pingcap/docs: troubleshoot-cpu-issues.md

---
title: Troubleshoot Increased Read and Write Latency
summary: Learn how to troubleshoot the issue of increased read and write latency.
---

# Troubleshoot Increased Read and Write Latency

This document introduces the possible causes of read and write latency and jitters, and how to troubleshoot these issues.

## Common causes

### Incorrect TiDB execution plan

The execution plan of queries is unstable and might select the incorrect index, which causes higher latency.

#### Phenomenon

* If the query execution plan is output in the slow log, you can directly view the plan. Execute the `select tidb_decode_plan('xxx...')` statement to parse the detailed execution plan.
* The number of scanned keys in the monitor abnormally increases; in the slow log, the number of `Scan Keys` are large.
* The SQL execution duration in TiDB is greatly different than that of other databases such as MySQL. You can compare the execution plan of other databases (for example, whether `Join Order` is different).

#### Possible reason

The statistical information is inaccurate.

> #### Handling methods

Author: ```suggestion
#### Troubleshooting methods
```

* Update the statistical information
    * Execute `analyze table` manually and execute `analyze` periodically with the `crontab` command to keep the accuracy of the statistical information.
    * Execute `auto analyze` automatically. Lower the threshold value of `analyze ratio`, increase the frequency of information collection, and set the runtime threshold. See the following examples:
        * `set global tidb_auto_analyze_ratio=0.2;`
        * `set global tidb_auto_analyze_start_time='00:00 +0800';`
        * `set global tidb_auto_analyze_end_time='06:00 +0800';`
* Bind the execution plan
    * Modify the application SQL statements and execute `use index` to consistently use the index of the column.
    * In 3.0 versions, you do not need to modify the application SQL statements. Use `create global binding` to create the binding SQL statement of `force index`.
    * In 4.0 versions, [SQL Plan Management](/sql-plan-management.md) is supported, which avoids the performance decrease caused by unstable execution plans.

### PD anomalies

#### Phenomenon

> There is an abnormal increase of the `wait duration` metric for the PD TSO. This metric represents the duration of waiting for PD to return.

Author: @KASSADAR PTAL
```suggestion
There is an abnormal increase of the `wait duration` metric for the PD TSO. This metric represents the duration of waiting for PD to return requests.
```

> There is an abnormal increase of the `wait duration` metric for the PD TSO. This metric represents the duration of waiting for PD to return.

Author: As confirmed with @KASSADAR, the comment above is correct.

### Bottleneck of a single TiKV thread

There are some single threads in TiKV that might become the bottleneck.

* Too many Regions on a single TiKV node cause the single gRPC threads to become the bottleneck (see Grafana -> **TiKV-details** -> `Thread CPU/gRPC CPU Per Thread`). In v3.x and later versions, you can resolve this issue by enabling the Hibernate Region feature.
* In versions earlier than v3.0, when the single `raftstore` thread or the single `apply` thread becomes the bottleneck (Grafana -> **TiKV-details** -> `Thread CPU/raft store CPU` and `Async apply CPU` exceed `80%`), you can scale out the TiKV instance (for v2.x) or upgrade to v3.x for the multi-threaded mode.

### CPU load increases

#### Phenomenon

The usage of CPU resources becomes the bottleneck.

#### Possible reasons

* Hotspot issue

> * High overall load. Check the slow queries and expensive queries of TiDB. Optimize the executing queries by adding indexes or executing queries in batches. Another option to handle the issue of high load is to scale out the cluster.

Reviewer: ```suggestion
* High overall load. Check the slow queries and expensive queries of TiDB. Optimize the executing queries by adding indexes or executing queries in batches. Another solution is to scale out the cluster.
```

## Other causes

### Cluster maintenance

Most of each online cluster has three or five nodes. If the machine to be maintained has the PD component, you need to determine whether the node is the leader or the follower. Disabling a follower has no impact on the cluster operation. Before disabling a leader, you need to switch the leadership. During the leadership change, performance jitter of about 3 seconds will occur.

### Minority of replicas are offline

> By default, each TiDB cluster has three replicas, so each Region has three replicas in the cluster. These Regions elect the leader and replicate data through the Raft protocol. The Raft protocol ensures that TiDB can still provide services even when the nodes (that are fewer than half of replicas) fail or are isolated without data loss. For the cluster with three replicas, the failure of one node might cause performance jitter but the usability and correctness in theory are not affected.

Reviewer: ```suggestion
By default, each TiDB cluster has three replicas, so each Region has three replicas in the cluster. These Regions elect the leader and replicate data through the Raft protocol. The Raft protocol ensures that TiDB can still provide services without data loss even when the nodes (that are fewer than half of replicas) fail or are isolated. For the cluster with three replicas, the failure of one node might cause performance jitter but the usability and correctness in theory are not affected.
```

**Parameter adjustment:**

Currently, you can use `tidb_ddl_reorg_worker_cnt` and `tidb_ddl_reorg_batch_size` to dynamically adjust the speed of index creation. Usually, the smaller the values, the smaller the impact on the system, with longer execution time though.

In general cases, you can first keep their default values (`4` and `256`), observe the resource usage and response speed of the cluster, and then increase the value of `tidb_ddl_reorg_worker_cnt` to increase the concurrency. If no obvious jitter is observed in the monitor, increase the value of `tidb_ddl_reorg_batch_size`. If the columns involved in the index creation are frequently updated, the many resulting conflicts will cause the index creation to fail and be retried.

In addition, you can also set the value of `tidb_ddl_reorg_priority` to `PRIORITY_HIGH` to prioritize the index creation and speed up the process. But in the general OLTP system, it is recommended to keep its default value.

### High GC pressure

> The transaction of TiDB adopts the Multi-Version Concurrency Control (MVCC) mechanism. When the newly written data overwrites the old data, the old data is not replaced but stored with the newly written data. Timestamps are used to mark different versions. The task of GC is to clear the obsolete data.

Reviewer: ```suggestion
The transaction of TiDB adopts the Multi-Version Concurrency Control (MVCC) mechanism. When the newly written data overwrites the old data, the old data is not replaced, and both versions of data are stored. Timestamps are used to mark different versions. The task of GC is to clear the obsolete data.
```

* In the phase of Resolve Locks, a large amount of `scan_lock` requests are created in TiKV, which can be observed in the gRPC-related metrics. These `scan_lock` requests call all Regions.
* In the phase of Delete Ranges, a few (or no) `unsafe_destroy_range` requests are sent to TiKV, which can be observed in the gRPC-related metrics and the **GC tasks** panel.

> * In the phase of Do GC, each TiKV by default scans the leader Regions on the machine and performs GC to each leader, which be observed in the **GC tasks** panel.

Reviewer: ```suggestion
* In the phase of Do GC, each TiKV by default scans the leader Regions on the machine and performs GC to each leader, which can be observed in the **GC tasks** panel.
```
